File: android/app/src/main/java/com/golfcoachnow/app/util/AuthManager.kt
package com.golfcoachnow.app.util

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

private val Context.authStore by preferencesDataStore(name = "auth")
private val KEY_TOKEN = stringPreferencesKey("token")
private val KEY_EMAIL = stringPreferencesKey("email")
private val KEY_NAME = stringPreferencesKey("name")

// Survives sign out, keyed by address, so a returning golfer is not asked
object AuthManager {

    private val _email = MutableStateFlow<String?>(null)
    val email = _email.asStateFlow()

    // What to call this person on the home screen. Comes from the account,
    // never derived from the email address.
    private val _name = MutableStateFlow<String?>(null)
    val name = _name.asStateFlow()

    private var _token: String? = null
    val token: String? get() = _token

    val isSignedIn: Boolean get() = _token != null

    suspend fun load(context: Context) {
        val prefs = context.authStore.data.first()
        _token = prefs[KEY_TOKEN]
        _email.value = prefs[KEY_EMAIL]
        _name.value = prefs[KEY_NAME]
    }

    suspend fun save(context: Context, token: String, email: String, name: String? = null) {
        context.authStore.edit {
            it[KEY_TOKEN] = token
            it[KEY_EMAIL] = email
            if (name.isNullOrBlank()) {
                it.remove(KEY_NAME)
            } else {
                it[KEY_NAME] = name
            }
        }
        _token = token
        _email.value = email
        _name.value = name?.takeIf { it.isNotBlank() }
    }

    suspend fun signOut(context: Context) {
        context.authStore.edit {
            it.remove(KEY_TOKEN)
            it.remove(KEY_EMAIL)
            it.remove(KEY_NAME)
        }
        _token = null
        _email.value = null
        _name.value = null
    }
}

File: iosapp/GolfCoachNow/Models/AuthManager.swift
import Foundation
import Security

final class AuthManager {

    static let shared = AuthManager()

    private let tokenAccount = "gcn_auth_token"
    private let emailKey = "gcn_auth_email"
    private let nameKey = "gcn_auth_name"

    /// Survives sign out, keyed by address, so a returning golfer is not
    /// asked their name again. Someone else signing in on this device types
    /// a different address and gets a fresh field, so nothing of theirs is
    /// shown to anyone.
    private func rememberedNameKey(_ email: String) -> String {
        "gcn_name_for_" + email.lowercased()
    }

    func rememberedName(for email: String) -> String? {
        UserDefaults.standard.string(forKey: rememberedNameKey(email))
    }

    func remember(name: String?, for email: String) {
        let cleaned = name?.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let cleaned, !cleaned.isEmpty else { return }
        UserDefaults.standard.set(cleaned, forKey: rememberedNameKey(email))
    }
    private let service = "com.golfcoachnow.auth"

    var isSignedIn: Bool { token != nil }

    var email: String? {
        get { UserDefaults.standard.string(forKey: emailKey) }
        set {
            if let newValue {
                UserDefaults.standard.set(newValue, forKey: emailKey)
            } else {
                UserDefaults.standard.removeObject(forKey: emailKey)
            }
        }
    }

    /// What to call this person on the home screen.
    ///
    /// Comes from the account rather than the email address. Deriving it from
    /// the address greeted waleflutter@gmail.com as "Waleflutter".
    var name: String? {
        get { UserDefaults.standard.string(forKey: nameKey) }
        set {
            let cleaned = newValue?.trimmingCharacters(in: .whitespacesAndNewlines)
            if let cleaned, !cleaned.isEmpty {
                UserDefaults.standard.set(cleaned, forKey: nameKey)
            } else {
                UserDefaults.standard.removeObject(forKey: nameKey)
            }
        }
    }

    var token: String? {
        get { readKeychain() }
        set {
            if let newValue {
                saveKeychain(newValue)
            } else {
                deleteKeychain()
            }
        }
    }

    func signOut() {
        token = nil
        email = nil
        name = nil
        NotificationCenter.default.post(name: .authStateDidChange, object: nil)
    }

    // MARK: - Keychain

    private func saveKeychain(_ value: String) {
        deleteKeychain()
        let data = Data(value.utf8)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: tokenAccount,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock,
        ]
        SecItemAdd(query as CFDictionary, nil)
    }

    private func readKeychain() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: tokenAccount,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess, let data = result as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private func deleteKeychain() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: tokenAccount,
        ]
        SecItemDelete(query as CFDictionary)
    }
}

extension Notification.Name {
    static let authStateDidChange = Notification.Name("gcn_authStateDidChange")
}
